### PR TITLE
add consul connection support for garmadon agent

### PR DESCRIPTION
**Pull request by mchataigner**
java agent could be started with: -javaagent:$PWD/garmadon-agent.jar=module1,module2-consul-FORWARDER_SERVICE_NAME

### CODE REVIEW

File: agent/src/main/java/com/criteo/hadoop/garmadon/agent/ConsulConnection.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+            LOGGER.debug("will use forwarder at " + host + ":" + port);
+        }
+
+        if (underlying != null) {
+            underlying.close();
+        }
+
+        underlying = new FixedConnection(host, port);
+
+        underlying.establishConnection();
```
Comment (ashangit) [inline agent/src/main/java/com/criteo/hadoop/garmadon/agent/ConsulConnection.java]: What happen if forwarder move to another host when underlying.establishConnection(), Does it loop for ever on FixedConnection.establishConnection?

File: agent/pom.xml [hunk 2]
```diff
@@ -190,6 +197,11 @@
                                     <pattern>org.slf4j</pattern>
                                     <shadedPattern>garmadon.org.slf4j</shadedPattern>
                                 </relocation>
+                                <!-- Relocating org.apache.http -->
```
[the remark this answers is not in the document]
Comment (mchataigner) [reply agent/pom.xml]: done

File: agent/pom.xml [hunk 3]
```diff
@@ -190,6 +197,11 @@
...
+                                <relocation>
+                                    <pattern>org.apache.http</pattern>
```
[the remark this answers is not in the document]
Comment (mchataigner) [reply agent/pom.xml]: yes

File: agent/pom.xml [hunk 4]
```diff
@@ -77,6 +77,13 @@
             <scope>provided</scope>
         </dependency>
 
+        <!-- discovery dependency -->
+        <dependency>
+            <groupId>com.ecwid.consul</groupId>
+            <artifactId>consul-api</artifactId>
+            <version>1.4.2</version>
```
[the remark this answers is not in the document]
Comment (mchataigner) [reply agent/pom.xml]: done

File: agent/pom.xml (hunk 3 above)
[the remark this answers is not in the document]
Comment (mchataigner) [reply agent/pom.xml]: nop, package is org.apache.http

File: agent/pom.xml [hunk 5]
```diff
@@ -190,6 +197,11 @@
                                     <pattern>org.slf4j</pattern>
                                     <shadedPattern>garmadon.org.slf4j</shadedPattern>
                                 </relocation>
+                                <!-- Relocating org.apache.http -->
+                                <relocation>
+                                    <pattern>org.apache.http</pattern>
+                                    <shadedPattern>shaded.org.apache.http</shadedPattern>
```
[the remark this answers is not in the document]
Comment (mchataigner) [reply agent/pom.xml]: same here

File: agent/pom.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (mchataigner) [reply agent/pom.xml]: done

File: agent/src/main/java/com/criteo/hadoop/garmadon/agent/ConsulConnection.java (hunk 1 above)
Comment (mchataigner) [reply agent/src/main/java/com/criteo/hadoop/garmadon/agent/ConsulConnection.java]: fixed

**Pull request merged**